contracts/transfer/src/state.rs

```diff
@@ -182,26 +292,113 @@ impl TransferState {
+    /// will have an inconsistent state.
+    ///
+    /// # Panics
+    /// Any failure in the checks performed in processing the transaction will
+    /// result in a panic. The contract expects the environment to roll back any
+    /// change in state.
+    ///
+    /// [`refund`]: [`TransferState::refund`]
+    fn spend_and_execute_moonlight(
+        &mut self,
+        tx: MoonlightTransaction,
+    ) -> Result<Vec<u8>, ContractError> {
+        transitory::put_transaction(tx);
+        let tx = transitory::unwrap_moonlight_tx();
+
+        // check the signature is valid and made by `from`
+        let payload = tx.payload();
+        let signature = *tx.signature();
+
+        let from = payload.from;
+        let digest = tx.signature_message();
+
+        if !rusk_abi::verify_bls(digest, from, signature) {
+            panic!("Invalid signature!");
+        }
+
+        // check `from` has the funds necessary to supress the total value
+        // available in this transaction, and that the `nonce` is higher than
+        // the currently held number. if these conditions are violated we panic
+        // since the transaction is invalid - either because the account doesn't
+        // have (enough) funds, or because they're possibly trying to reuse a
+        // previously used signature (i.e. a replay attack).
+        //
+        // Afterwards be simply deduct the total amount of the transaction from
+        // the balance, increment the nonce, and rely on `refund` to be called
+        // after a successful exit.
+        let from_bytes = from.to_bytes(); // TODO: this is expensive. maybe we should address the
+                                          //       fact that `AccountPublicKey` doesn't impl `Ord`
+                                          //       so we can just use it directly as a key in the
+                                          //       `BTreeMap`
+
+        // the total value carried by a transaction is the sum of the value, the
+        // deposit, and gas limit * gas_price.
+        let total_value = payload.value
+            + payload.deposit
+            + payload.gas_limit * payload.gas_price;
+
+        match self.accounts.get_mut(&from_bytes) {
+            Some(account) => {
+                if total_value > account.balance {
+                    panic!("Account doesn't have enough funds");
+                }
+                if payload.nonce <= account.nonce {
+                    panic!("Replayed nonce");
                 }
+
+                account.balance -= total_value;
+                account.nonce = payload.nonce;
+            }
+            None => panic!("Account has no funds"),
+        }
+
+        // if there is a value carried by the transaction but no key specified
+        // in the `to` field, we just give the value back to `from`.
+        if payload.value > 0 {
+            let key = match payload.to {
+                Some(to) => to.to_bytes(),
+                None => from_bytes,
```

Comment:
why is the `to` key set to `from_bytes`, the serialized `from`, when `to` is not found? Does that mean we send the funds back to `from` when `to` is not found?

Reply:
If the transaction does not specify the account the funds will go to, then it simply goes back to the origin.